### PR TITLE
pkg/trace/api: ensure http errors get logged

### PR DESCRIPTION
**Pull request by gbbr**
This change adds an error logger to the HTTP receiver API. Previously, HTTP errors from the standard library were being missed in our logs (such as rate-limiting when hitting the connection limit) and was causing unnecessary support tickets and debugging difficulties.

### CODE REVIEW

File: pkg/trace/api/api.go [hunk 1]
```diff
@@ -442,3 +444,11 @@ func tracesFromSpans(spans []pb.Span) pb.Traces {
...
 	return traces
 }
+
+// logWriter is an error logging function which can be written to.
+type logWriter func(v ...interface{}) error
+
+// Write implements io.Writer
+func (lw logWriter) Write(p []byte) (n int, err error) {
+	return len(p), lw(string(p))
+}
```
Comment (AlexJF) [inline pkg/trace/api/api.go]: This is a bit weird x)

File: pkg/trace/api/api.go [hunk 2]
```diff
@@ -140,6 +141,7 @@ func (r *HTTPReceiver) Listen(addr, logExtra string) error {
 	r.server = &http.Server{
 		ReadTimeout:  timeout,
 		WriteTimeout: timeout,
+		ErrorLog:     fmtlog.New(logWriter(log.Error), "http.Server: ", 0),
```
Comment (AlexJF) [inline pkg/trace/api/api.go]: Will this use the agent logger with its throttling mechanism?

Comment (gbbr) [reply pkg/trace/api/api.go]: I'd say so, given that it forwards all calls to `log.Error`, which uses the underlying logger set up with throttling.

File: pkg/trace/api/api.go (hunk 1 above)
Comment (gbbr) [reply pkg/trace/api/api.go]: Which part do you mean? To me it's weird that the first return value might not be 100% accurate but I can't think of a better solution. I'm open to suggestions!

Comment (AlexJF) [reply pkg/trace/api/api.go]: I mean the whole thing about having a type that is aliased to a function and calling `lw(...)` directly from a method on that type/function. Then the part where we do `logWriter(log.Error)` which is actually doing a type conversion not an invocation.

I see how/why it works after looking at it carefully but initially I got a bit confused as to what was going on.

**Review by AlexJF — DISMISSED** [review]

File: pkg/trace/api/api.go (hunk 2 above)
Comment (AlexJF) [reply pkg/trace/api/api.go]: Right got a bit mixed up with the `fmtlog` vs `log` packages :+1:

File: pkg/trace/api/api.go (hunk 1 above)
Comment (gbbr) [reply pkg/trace/api/api.go]: Initially I had a struct but then I thought of [`http.HandlerFunc`](http://golang.org/pkg/net/http#HandlerFunc) which is somewhat similar and went with that. Maybe `funcWriter` would've been better but 🤷‍♂️

File: pkg/trace/api/api.go (hunk 2 above)
Comment (gbbr) [reply pkg/trace/api/api.go]: No wonder. I've changed the names of that in #3134. All seelog stuff will be `seelog.*`

File: pkg/trace/api/api.go (hunk 1 above)
Comment (gbbr) [reply pkg/trace/api/api.go]: I had to rebase on a new `master` and made this a bit better (I think) both in terms of naming and error returns. Can you PTAL? Your review became stale as a result of the rebase.

**Review by AlexJF — APPROVED** [review]

**Pull request merged**